implementation/src/main/java/io/smallrye/config/ConfigSourceMap.java

```diff
@@ -0,0 +1,180 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.smallrye.config;
+
+import java.util.AbstractCollection;
+import java.util.AbstractMap;
+import java.util.AbstractSet;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.BiConsumer;
+
+import org.eclipse.microprofile.config.spi.ConfigSource;
+
+/**
+ * A {@link Map Map&lt;String, String>} which is backed by a {@link ConfigSource}.
+ * This should not be used to implement {@link ConfigSource#getProperties()} on {@code ConfigSource}
+ * instances which do not override {@code getPropertyNames()}, as this will result in infinite recursion.
+ *
+ * @implNote The key set of the map is the result of calling {@link ConfigSource#getPropertyNames()}; the rest
+ * of the map operations are derived from this method and {@link ConfigSource#getValue(String)}.
+ * The values collection and entry set are instantiated lazily and cached.
+ * The implementation attempts to make no assumptions about the efficiency of the backing implementation and
+ * prefers the most direct access possible.
+ * <p>
+ * Some mutate operations are supported (specifically removals), where they can be performed on the backing key set, if
+ * the backing implementation allows them.
+ * Keys cannot be added to the map, nor values assigned, under any circumstances.
+ */
+public class ConfigSourceMap extends AbstractMap<String, String> implements Map<String, String> {
+    private final ConfigSource delegate;
+    private Values values;
+    private EntrySet entrySet;
+
+    /**
+     * Construct a new instance.
+     *
+     * @param delegate the delegate configuration source (must not be {@code null})
+     */
+    public ConfigSourceMap(final ConfigSource delegate) {
+        this.delegate = Objects.requireNonNull(delegate, "delegate must not be null");
+    }
+
+    public int size() {
+        return delegate.getPropertyNames().size();
+    }
+
+    public boolean isEmpty() {
+        // may be cheaper in some cases
+        return delegate.getPropertyNames().isEmpty();
+    }
+
+    public boolean containsKey(final Object key) {
+        //noinspection SuspiciousMethodCalls - it's OK in this case
+        return delegate.getPropertyNames().contains(key);
+    }
+
+    public String get(final Object key) {
+        return key instanceof String ? delegate.getValue((String) key) : null;
+    }
+
+    public Set<String> keySet() {
+        return delegate.getPropertyNames();
+    }
+
+    public void clear() {
+        delegate.getPropertyNames().clear();
+    }
+
+    public String remove(final Object key) {
+        final String value = get(key);
+        //noinspection SuspiciousMethodCalls - it's OK in this case
+        return delegate.getPropertyNames().remove(key) ? value : null;
```

Comment:
`remove` could throw a `UnsupportedOperationException` if the list is unmodifiable